astropy/io/fits/hdu/table.py

```diff
@@ -165,8 +165,9 @@ def _get_tbdata(self):
             # We have a heap; include it in the raw_data
             raw_data = self._get_raw_data(self._data_size, np.uint8,
                                           self._data_offset)
-            data = raw_data[:self._theap].view(dtype=columns.dtype,
-                                               type=np.rec.recarray)
+            tbsize = self._header['NAXIS1'] * self._header['NAXIS2']
```

Comment:
It occurs to me there's a bunch of places in this file where this calculation is being made.  It's cheap, so that's not a problem.  But I think it's a bit unfortunate that it's so dependent on values in the headers (which users sometimes break; see #3836).

There should probably some way to better manage the actual size of table data somewhere.